src/System.Net.Http/src/System/Net/Http/Headers/KnownHeader.cs

```diff
@@ -12,34 +13,36 @@ internal sealed class KnownHeader
         private readonly HttpHeaderType _headerType;
         private readonly HttpHeaderParser _parser;
         private readonly string[] _knownValues;
+        private readonly byte[] _asciiBytesWithColonSpace;
 
-        public KnownHeader(string name, HttpHeaderType headerType, HttpHeaderParser parser, string[] knownValues = null)
+        public KnownHeader(string name) : this(name, HttpHeaderType.Custom, null)
         {
             Debug.Assert(!string.IsNullOrEmpty(name));
             Debug.Assert(HttpRuleParser.GetTokenLength(name, 0) == name.Length);
-            Debug.Assert(headerType != HttpHeaderType.Custom);
-            Debug.Assert(parser != null);
-
-            _name = name;
-            _headerType = headerType;
-            _parser = parser;
-            _knownValues = knownValues;
         }
 
-        public KnownHeader(string name)
+        public KnownHeader(string name, HttpHeaderType headerType, HttpHeaderParser parser, string[] knownValues = null)
         {
             Debug.Assert(!string.IsNullOrEmpty(name));
             Debug.Assert(HttpRuleParser.GetTokenLength(name, 0) == name.Length);
 
             _name = name;
-            _headerType = HttpHeaderType.Custom;
-            _parser = null;
+            _headerType = headerType;
+            _parser = parser;
+            _knownValues = knownValues;
+
+            _asciiBytesWithColonSpace = new byte[name.Length + 2]; // + 2 for ':' and ' '
+            int asciiBytes = Encoding.ASCII.GetBytes(name, _asciiBytesWithColonSpace);
+            Debug.Assert(asciiBytes == name.Length);
+            _asciiBytesWithColonSpace[_asciiBytesWithColonSpace.Length - 2] = (byte)':';
+            _asciiBytesWithColonSpace[_asciiBytesWithColonSpace.Length - 1] = (byte)' ';
         }
 
         public string Name => _name;
         public HttpHeaderParser Parser => _parser;
         public HttpHeaderType HeaderType => _headerType;
         public string[] KnownValues => _knownValues;
+        public byte[] AsciiBytesWithColonSpace => _asciiBytesWithColonSpace;
```

Comment:
Should this be a ReadOnlySpan<byte>?  Just to prevent accidental mutation.

Reply:
Making it a `ReadOnlySpan<byte>` could add some cost, as we can't pass that to an asynchronous operation, so if the buffer isn't large enough to fit it when we go to write it, we'd have to allocate.  We could make it a `ReadOnlyMemory<T>`, though that'll add a small amount of overhead to extract the byte array.  I agree it'd be nice to have the accidental mutation benefit, but we already have other mutable things on this type (e.g. KnownValues), so I'm not sure it's really worthwhile.

Reply:
Ok